RFCs/RFC-001: Introspect streams.md

```diff
@@ -0,0 +1,62 @@
+# RFC-001: Introspect Streams
+
+**To be Reviewed By:** 11/19/2019
+
+**Authors:** Swapnil Bawaskar
+
+**Status:** **Active** | Accepted | Dropped | Superseded
+
+**Superseded by:** N/A
+
+**Related:** N/A
+
+
+## Problem
+When the streaming runtime is installed, riff users can create streams (backed by a stream provider e.g. kafka). Users can write functions that read from and write to these streams,
+but there is no easy mechanism to get events into the stream and look into the contents of the stream which are helpful while developing/demoing.
+
+### Anti-Goals
+We will only address this problem for development/demos, not production, so topics like auth/authz are out of scope for this document.
+
+## Solution
+### Ideal solution:
+The  is to provide users with only one cli (riff cli) that will enable users to send/receive messages to/from a stream (commands below). We will have to ask the users to port-forward the gateway service if they are running outside the cluster.
+
+#### User Impact
+We should introduce the following commands to the riff cli:
+- send a massage to the stream:
+
+    `riff streaming stream publish <stream-name> --payload <payload-as-string> --headers <headers-as-string>`  
+    content type has already been specified while creating the stream.
+
+- subscribe for messages from a stream:
+
+    `riff streaming stream subscribe <stream-name> --offset <long-offset>`  
+    this will display all the message from the given offset and then block to display further messages, essentially preventing user from entering subsequent commands.
+    The messages will be displayed as json with the following fields:
+    ``` 
+    {
+        "payload": "the users payload",
```

Comment:
Why not use JSON (it allows piping in jq for example), but should then clarify how the payload is displayed. The raw payload is bytes. How should that be displayed in JSON? Should we always output the whole payload?

Reply:
the payload would need to be base64 encoded in order to be encoded as a string in JSON. Or an array of numbers where each item in the array is a byte (big or little endian?).

Reply:
added another option `--payload-file <relative-filepath>` to the `publish` command. These will be base64 encoded so that subscribe will always be able to display strings.

Reply:
Why would you base64 encode the content of the file? At that point bytes is fine. The issue Eric raised above is that you can't have arbitrary bytes in JSON. Remove JSON, remove the binary encoding issue.

Reply:
I think we should keep JSON, so that users that users can then process the output if desired. 

In order to display the payload as string in json, I was hoping to ensure that the payload is always published as a string. Hence, if you had binary content in a file, base64 encode it. I now realize that users functions will be publishing to the stream as well, so encoding while publishing will not work.